File: 0145_binary_tree_postorder_traversal.py
# n: node_count
# Time: O(n)
# Space: O(n)
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        val_list = []
        self.dfs(root, val_list)
        return val_list

    def dfs(
            self, root: Optional[TreeNode], val_list: List[TreeNode]
    ) -> List[int]:
        if not root:
            return

        self.dfs(root.left, val_list)
        self.dfs(root.right, val_list)
        val_list.append(root.val)

File: go/0145_binary_tree_postorder_traversal.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// n = number of nodes in the tree
// Time:  O(n)
// Space: O(n)
func postorderTraversal(root *TreeNode) []int {
	result := []int{}
	var postorder func(*TreeNode)

	postorder = func(root *TreeNode) {
		if root == nil {
			return
		}

		postorder(root.Left)
		postorder(root.Right)
		result = append(result, root.Val)
	}

	postorder(root)

	return result
}
